langkit/templates/pkg_rewriting_impl_body_ada.mako

```diff
@@ -27,136 +27,120 @@ package body ${ada_lib_name}.Rewriting_Implementation is
-
-   procedure Pre_Check_No_Diagnostics (Unit : Internal_Unit) is
-   begin
-      if Has_Diagnostics (Unit) then
-         Pre_Fail ("Cannot retrieve the handle of a node belonging to an " &
-                   "unit that has diagnostics");
-      end if;
-   end;
-   -- Check that the given analysis unit does not have any diagnostics.
-
-   procedure Pre_Check_Is_Token_Kind (Kind : ${root_node_kind_name}) is
-   begin
-      if not Is_Token_Node (Kind) then
-         Pre_Fail ("Expected a token node. Got " & Kind'Image);
-      end if;
-   end Pre_Check_Is_Token_Kind;
-   -- Check that the given kind corresponds to a token node kind
-
-   procedure Pre_Check_Is_Not_Token_Kind (Kind : ${root_node_kind_name}) is
-   begin
-      if Is_Token_Node (Kind) then
-         Pre_Fail ("Got unexpected token node " & Kind'Image);
-      end if;
-   end Pre_Check_Is_Not_Token_Kind;
-   -- Check that the given kind does *not* correspond to a token node kind
-
-   procedure Pre_Check_Is_List_Kind (Kind : ${root_node_kind_name}) is
-   begin
-      if not Is_List_Node (Kind) then
-         Pre_Fail ("Expected a list node. Got " & Kind'Image);
-      end if;
-   end Pre_Check_Is_List_Kind;
-   -- Check that the given kind corresponds to a list node kind
-
-   procedure Pre_Check_Null_Or_Untied
-     (Handle     : Node_Rewriting_Handle) is
-   begin
-      if Handle /= No_Node_Rewriting_Handle and then Tied (Handle) then
-         Pre_Fail ("The given node must not be tied to another " &
-                   "rewriting context.");
-      end if;
-   end Pre_Check_Null_Or_Untied;
-   -- Check that the given node rewriting handle is not tied to any unit
-   -- rewriting handle.
-   -- Note: Null matches this predicate.
+   ---------------
+   -- PRE_CHECK --
+   ---------------
 
-   procedure Pre_Check_Tied
-     (Handle     : Node_Rewriting_Handle) is
-   begin
-      if not Tied (Handle) then
-         Pre_Fail ("The given node must be tied to an analysis unit.");
-      end if;
-   end Pre_Check_Tied;
-   -- Check that the given node rewriting handle is well tied to an analysis
-   -- unit.
+   procedure Pre_Check (Value : Boolean; Msg : String);
+   --  Raise a Precondition_Failure exception with the given message
+   --  if the Value is False.
 
-   procedure Pre_Check_Null_Or_Valid_Context
-     (Node : Node_Rewriting_Handle;
-      Handle : Rewriting_Handle) is
+   procedure Pre_Check (Value : Boolean; Msg : String) is
    begin
-      if Node /= No_Node_Rewriting_Handle and then Context (Node) = Handle then
-         Pre_Fail ("The given node has an invalid context.");
+      if not Value then
+         raise Precondition_Failure with Msg;
       end if;
-   end Pre_Check_Null_Or_Valid_Context;
-   -- Check that the node rewriting handle is either null, or belongs to the
-   -- given rewriting handle.
+   end Pre_Check;
+
+   <%def name="pre_check_rw_handle(handle)">
+      --  Check that rewriting handle ${handle} is not null
```

Comment:
These comments document the macro itself, but will clutter the generated code. Can you turn them into Mako comments? (## Check that […])